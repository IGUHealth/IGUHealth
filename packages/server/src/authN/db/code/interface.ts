import * as s from "zapatos/schema";

import { ModelManagement } from "../interface.js";
import { AuthorizationCode } from "./types.js";

export interface AuthorizationCodeManagement
  extends ModelManagement<
    AuthorizationCode,
    s.authorization_code.Whereable,
    s.authorization_code.Insertable,
    s.authorization_code.Updatable
  > {}
