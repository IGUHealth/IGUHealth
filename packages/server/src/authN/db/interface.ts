import { KoaExtensions } from "../../fhir-api/types.js";

export interface ModelManagement<
  Model extends Insertable & Updateable,
  Whereable,
  Insertable,
  Updateable,
> {
  get(
    ctx: KoaExtensions.IGUHealthServices["iguhealth"],
    id: string,
  ): Promise<Model | undefined>;
  search(
    ctx: KoaExtensions.IGUHealthServices["iguhealth"],
    where: Whereable,
  ): Promise<Model[]>;
  create(
    ctx: KoaExtensions.IGUHealthServices["iguhealth"],
    model: Insertable,
  ): Promise<Model>;
  update(
    ctx: KoaExtensions.IGUHealthServices["iguhealth"],
    id: string,
    update: Updateable,
  ): Promise<Model>;
  delete(
    ctx: KoaExtensions.IGUHealthServices["iguhealth"],
    where: Whereable,
  ): Promise<void>;
}
