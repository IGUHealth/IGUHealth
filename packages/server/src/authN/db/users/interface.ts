import * as db from "zapatos/db";
import * as s from "zapatos/schema";

import { TenantClaim } from "@iguhealth/jwt";

import { KoaContext } from "../../../fhir-api/types.js";
import { ModelManagement } from "../interface.js";
import { LoginParameters, User } from "./types.js";

export interface UserManagement
  extends ModelManagement<
    User,
    s.users.Whereable,
    s.users.Insertable,
    s.users.Updatable
  > {
  login<T extends keyof LoginParameters>(
    ctx: KoaContext.FHIRServices,
    type: T,
    parameters: LoginParameters[T],
  ): Promise<User | undefined>;
  getTenantClaims(
    ctx: KoaContext.FHIRServices,
    id: string,
  ): Promise<TenantClaim<s.user_role>[]>;
}
