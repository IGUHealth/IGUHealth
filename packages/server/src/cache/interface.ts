import type { Tenant } from "../fhir/context.js";

export interface IOCache<CTX extends { tenant: Tenant }> {
  get(ctx: CTX, key: string): Promise<string | number | null>;
  set(ctx: CTX, key: string, value: string | number): Promise<void>;
}
