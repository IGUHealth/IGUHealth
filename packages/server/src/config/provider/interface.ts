import type { ConfigSchema } from "../../json-schemas/schemas/config.schema.js";

export interface ConfigProvider {
  get<K extends keyof ConfigSchema>(key: K): ConfigSchema[K];
}
