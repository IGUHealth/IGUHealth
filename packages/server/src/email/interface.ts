import { HTTPResponse } from "../fhir-http/index.js";

export type Email = {
  to: string;
  from: string;
  subject?: string;
} & EmailContent;

type EmailContent = { text: string } | { html: string };

export interface EmailProvider {
  sendEmail: (email: Email) => Promise<HTTPResponse>;
}
