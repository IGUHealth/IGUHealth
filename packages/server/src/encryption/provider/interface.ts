export interface EncryptionProvider {
  encrypt(context: Record<string, string>, data: string): Promise<string>;
  decrypt(context: Record<string, string>, data: string): Promise<string>;
}
