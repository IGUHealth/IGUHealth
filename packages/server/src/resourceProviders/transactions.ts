import graphlib from "@dagrejs/graphlib";
import pg from "pg";

import { evaluateWithMeta } from "@iguhealth/fhirpath";
import { Bundle, Reference, uri } from "@iguhealth/fhir-types/r4/types";
import { OperationError, outcomeFatal } from "@iguhealth/operation-outcomes";
import { FHIRServerCTX } from "../fhir/types.js";

function getTransactionFullUrls(transaction: Bundle): Record<string, number> {
  const record: Record<string, number> = {};

  for (const idx in transaction.entry) {
    const entry = transaction.entry[parseInt(idx)];
    if (entry.fullUrl) {
      record[entry.fullUrl] = parseInt(idx);
    }
  }
  return record;
}

type LocationsToUpdate = {
  [key: string]: (string | number)[][] | undefined;
};

export function buildTransactionTopologicalGraph(
  ctx: FHIRServerCTX,
  transaction: Bundle
): {
  order: string[];
  locationsToUpdate: LocationsToUpdate;
  graph: graphlib.Graph;
} {
  const entries = transaction.entry || [];
  const graph = new graphlib.Graph({ directed: true });
  const urlToIndice = getTransactionFullUrls(transaction);
  const locationsToUpdate: LocationsToUpdate = {};

  for (const idx in entries) {
    const entry = entries[parseInt(idx)];
    graph.setNode(idx, entry.fullUrl);

    // Pull dependencies from references.
    if (entry.resource) {
      const resourceReferences = evaluateWithMeta(
        "$this.descendants().ofType(Reference)",
        entry.resource,
        {
          meta: {
            type: entry.resource.resourceType as uri,
            getSD: (type) => {
              const canonical = ctx.resolveTypeToCanonical(type);
              if (!canonical)
                throw new OperationError(
                  outcomeFatal(
                    "exception",
                    `Could not resolve canonical for type '${type}'.`
                  )
                );
              return ctx.resolveCanonical("StructureDefinition", canonical);
            },
          },
        }
      );

      const bundleDependencies = resourceReferences.filter((v) => {
        const ref = (v.valueOf() as Reference).reference;
        if (ref && urlToIndice[ref] !== undefined) return true;
        return false;
      });

      for (const dep of bundleDependencies) {
        const ref = (dep.valueOf() as Reference).reference as string;
        locationsToUpdate[ref] = locationsToUpdate[ref] || [];
        const location = dep.location();
        if (!location)
          throw new OperationError(
            outcomeFatal(
              "exception",
              "Transaction processing could not find dependents location."
            )
          );

        const bundleEntryLocation = ["entry", idx, "resource", ...location];
        (locationsToUpdate[ref] || []).push(bundleEntryLocation);
        graph.setEdge(urlToIndice[ref as string].toString(), idx);
      }
    }
  }
  if (!graphlib.alg.isAcyclic(graph)) {
    const cycles = graphlib.alg.findCycles(graph);
    throw new OperationError(
      outcomeFatal(
        "exception",
        `Transaction bundle has cycles at following indice paths ${cycles
          .map((cycle) => cycle.join("->"))
          .join(", ")}.`
      )
    );
  }
  return { locationsToUpdate, graph, order: graphlib.alg.topsort(graph) };
}

async function retryFailedTransactions<ReturnType>(
  ctx: FHIRServerCTX,
  numberOfRetries: number,
  execute: () => Promise<ReturnType>
): Promise<ReturnType> {
  for (let i = 0; i < numberOfRetries; i++) {
    try {
      const res = await execute();
      return res;
    } catch (e) {
      if (!(e instanceof pg.DatabaseError)) {
        ctx.logger.error("Error during transaction:", e);
        throw e;
      }
      // Only going to retry on failed transactions
      if (e.code !== "40001") {
        ctx.logger.error("Postgres error threw :", e);
        throw e;
      }
      ctx.logger.warn(`Retrying transaction : ${i + 1}`, { retry: i + 1 });
    }
  }

  ctx.logger.error(`Max number of retries exceeded '${numberOfRetries}'`);
  throw new OperationError(
    outcomeFatal(
      "lock-error",
      `Could not apply transaction after '${numberOfRetries}' retries.`
    )
  );
}

/*
 ** Postgres transactional isolation levels.
 */
export enum ISOLATION_LEVEL {
  /*
   ** Dirty Read:            Not possible
   ** Nonrepeatable Read:	 Possible
   ** Phantom Read:          Possible
   ** Serialization Anomaly: Possible
   */
  ReadCommitted = 0,
  /*
   ** Dirty Read:            Not possible
   ** Nonrepeatable Read:	 Not possible
   ** Phantom Read:          Allowed, but not in PG
   ** Serialization Anomaly: Possible
   */
  RepeatableRead = 1,
  /*
   ** Dirty Read:            Not possible
   ** Nonrepeatable Read:	 Not possible
   ** Phantom Read:          Not possible
   ** Serialization Anomaly: Not possible
   */
  Serializable = 2,
}

function begin(isolation_level: ISOLATION_LEVEL) {
  switch (isolation_level) {
    case ISOLATION_LEVEL.ReadCommitted: {
      return "BEGIN";
    }
    case ISOLATION_LEVEL.RepeatableRead: {
      return "BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ";
    }
    case ISOLATION_LEVEL.Serializable: {
      return "BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE";
    }
    default: {
      throw new OperationError(
        outcomeFatal("exception", `Invalid isolation level ${isolation_level}`)
      );
    }
  }
}

export async function transaction<T>(
  isolation_level: ISOLATION_LEVEL,
  ctx: FHIRServerCTX,

  client: pg.PoolClient,
  body: (ctx: FHIRServerCTX) => Promise<T>
): Promise<T> {
  if (ctx.inTransaction) return body(ctx);
  return retryFailedTransactions(ctx, 5, async () => {
    try {
      await client.query(begin(isolation_level));
      const returnV = await body({ ...ctx, inTransaction: true });
      await client.query("END");
      return returnV;
    } catch (e) {
      await client.query("ROLLBACK");
      throw e;
    }
  });
}
