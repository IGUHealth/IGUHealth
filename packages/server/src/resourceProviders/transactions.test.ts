import path from "path";
import { test, expect } from "@jest/globals";

import {
  ResourceType,
  Resource,
  StructureDefinition,
} from "@iguhealth/fhir-types/r4/types";
import { loadArtifacts } from "@iguhealth/artifacts";
import { FHIRClientSync } from "@iguhealth/client/interface";
import { OperationError, outcomeFatal } from "@iguhealth/operation-outcomes";

import MemoryDatabase from "./memory/sync.js";
import { buildTransactionTopologicalGraph } from "./transactions";
import { testServices } from "./test_ctx.js";
import { FHIRServerCTX } from "../fhirServer.js";

function createMemoryDatabase(
  resourceTypes: ResourceType[]
): FHIRClientSync<unknown> {
  const database = MemoryDatabase<unknown>({});
  const artifactResources: Resource[] = resourceTypes
    .map((resourceType) =>
      loadArtifacts(resourceType, path.join(__dirname, "../"), true)
    )
    .flat();
  for (const resource of artifactResources) {
    database.create({}, resource);
  }
  return database;
}

const memDB = createMemoryDatabase(["StructureDefinition"]);

function getSD(type: string): StructureDefinition | undefined {
  return memDB.read({}, "StructureDefinition", type);
}

const CTX = {
  ...testServices,
  resolveSD: (ctx: FHIRServerCTX, type: string) => getSD(type),
};

test("Generate a graph from a transaction", () => {
  const result = buildTransactionTopologicalGraph(CTX, {
    resourceType: "Bundle",
    type: "transaction",
    entry: [
      {
        fullUrl: "urn:oid:2",
        resource: {
          resourceType: "Patient",
          generalPractitioner: [{ reference: "urn:oid:1" }],
        },
      },
      {
        fullUrl: "urn:oid:1",
        resource: {
          resourceType: "Practitioner",
          name: [{ given: ["Bob"] }],
        },
      },
    ],
  });
  expect(result).toMatchSnapshot();
  expect(result.order).toEqual(["1", "0"]);
});

test("Test Cyclical", () => {
  expect(() => {
    return buildTransactionTopologicalGraph(CTX, {
      resourceType: "Bundle",
      type: "transaction",
      entry: [
        {
          fullUrl: "urn:oid:2",
          resource: {
            resourceType: "Patient",
            generalPractitioner: [{ reference: "urn:oid:1" }],
          },
        },
        {
          fullUrl: "urn:oid:1",
          resource: {
            extension: [
              { url: "test", valueReference: { reference: "urn:oid:2" } },
            ],
            resourceType: "Practitioner",
            name: [{ given: ["Bob"] }],
          },
        },
      ],
    });
  }).toThrow(
    new OperationError(
      outcomeFatal(
        "exception",
        `Transaction bundle has cycles at following indices ${JSON.stringify(
          []
        )}.`
      )
    )
  );
  try {
    buildTransactionTopologicalGraph(CTX, {
      resourceType: "Bundle",
      type: "transaction",
      entry: [
        {
          fullUrl: "urn:oid:2",
          resource: {
            resourceType: "Patient",
            generalPractitioner: [{ reference: "urn:oid:1" }],
          },
        },
        {
          fullUrl: "urn:oid:1",
          resource: {
            extension: [
              { url: "test", valueReference: { reference: "urn:oid:2" } },
            ],
            resourceType: "Practitioner",
            name: [{ given: ["Bob"] }],
          },
        },
      ],
    });
  } catch (e) {
    if (e instanceof OperationError) {
      expect(e.operationOutcome.issue).toEqual([
        {
          code: "exception",
          diagnostics:
            "Transaction bundle has cycles at following indice paths 1->0.",
          expression: undefined,
          severity: "fatal",
        },
      ]);
    } else {
      throw e;
    }
  }
});
