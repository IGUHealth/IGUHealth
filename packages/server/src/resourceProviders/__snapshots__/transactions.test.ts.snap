// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate a graph from a transaction 1`] = `
{
  "graph": Graph {},
  "locationsToUpdate": {
    "urn:oid:1": [
      [
        "entry",
        "resource",
        "0",
        "generalPractitioner",
        0,
      ],
    ],
  },
  "order": [
    "1",
    "0",
  ],
}
`;
